Add SWMR functionality to filestore for AreaDetector HDF5

stuwilkins opened the pull request:
This is a WIP (DO NOT MERGE) to add SWMR functionality.

stuwilkins commented:
Attn @danielballan, @ambarb, @cmazzoli 

This is now working for counts where there are multiple events:

See this [notebook](https://nbviewer.jupyter.org/gist/stuwilkins/b9c9cb1505a891b0ea87faf2fd3107c9) for an example of reading data while the scan is running

danielballan commented:
Looks good. Let's generate SWMR data with a new SPEC, `AD_HDF5_SMWR`. I guess it's fine to use one general-purpose handler that opens either `AD_HDF5` or `AD_HDF5_SWMR`, but it might be useful to distinguish which is which at some point in the future (without opening the file to look).

tacaswell commented:
This seems reasonable to me.

stuwilkins commented:
@danielballan I think that makes a lot of sense. Given that SWMR is an option at creation time of the data then I think is is a really good idea to write the entry with the correct SPEC. The bluesky could even be intelligent enough to see if SWMR is on and encode the spec with AD_HDF5_SWMR or otherwise.

danielballan commented:
Makes sense to me.

stuwilkins commented:
@danielballan @tacaswell 

I think this is now good to go, I switched up a bit the HDF5 classes and added the SWMR options. I also added the AreaDetector timestamp handlers as these are generic to AD and not CSX special

Also, unit tests are written, but tests for SWMR are disabled until updated h5py and libhdf5 are done. How do you want to handle it?

So... comments or are we good to go?

tacaswell commented on filestore/handlers.py:
```diff
@@ -109,15 +95,32 @@ def __init__(self, filename, key, frame_per_point=1):
         self._key = key
         self._file = None
         self._dataset = None
+        self._swmr = False
```
where does `self._swmr` ever get set to `True`?

stuwilkins replied:
Fixed. Now at line 169.

stuwilkins commented on filestore/handlers.py:
```diff
@@ -142,7 +140,109 @@ def __init__(self, filename, frame_per_point=1):
+        self._file = h5py.File(self._filename, 'r', swmr=True)
+
+    def __call__(self, point_number):
+        if self._dataset is not None:
```
@tacaswell I think this might be better, it just uses the class override rather than using a variable ``self._swmr``

tacaswell commented:
That looks better to me.

danielballan commented:
Looks good to me. I agree the tests can wait until we are deploying latest stable hdf5 lib everywhere. It's worth the effort to wrangle that right now. Merging.

stuwilkins commented:
@danielballan @tacaswell thanks!
